server/src/main/java/org/elasticsearch/http/AbstractHttpServerTransport.java

```diff
@@ -247,20 +252,27 @@ protected void doStop() {
                 }
             }
         }
-        gracefullyCloseConnections();
+        shuttingDown = true;
         refCounted.decRef();
         boolean closed = false;
+
+        // httpChannels has no more puts after shuttingDown because of the guard in serverAcceptedChannel
+        httpChannels.values().forEach(RequestTrackingHttpChannel::setCloseWhenIdle);
```

Comment:
A possible race condition?
* accepted new channel before stop called
* closewhenidle called here before the incomingRequest is able to increment the refcounted, so it gets closed

Reply:
So setCloseWhenIdle would be called here, the refCounted will be decremented to 0, and the closer listener will be called. The check for null tracking channel in incomingRequest is meant to account for this case, right?

It's unfortunate we can't distinguish between this edge case and a real internal error case (called incomingRequest without serverAcceptedChannel). I wonder if we could keep a placeholder in the map so we don't lose track that the channel did exist. Additionally, it seems like there could still be a race condition, in that the get call in incomingRequest could occur before or after the channel is actually removed. Perhaps there should be a hasReferences() check on the tracking channel's refcounted? If the tracking channel were kept around in the map, that could allow accounting for this edge case. But even then, it seems like without some locking there will always be a race condition?

Reply:
> The check for null tracking channel in incomingRequest is meant to account for this case, right?

This was actually exposed via the final test in  [SecurityNetty4HttpServerTransportTests.testMalformedRequestDispatchedNoAuthn](https://github.com/elastic/elasticsearch/blob/0e95bb4ef7ffd39aa7f88526df0f7c05ce2138c1/x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SecurityNetty4HttpServerTransportTests.java#L676)